ultravox/tools/infer_tool.py

```diff
@@ -190,25 +192,46 @@ def dataset_infer(inference: base.VoiceInference, args: InferArgs):
     if args.seed is not None:
         ds_args.shuffle_seed = args.seed
     ds = datasets.create_dataset(args.data_sets[0], ds_args)
-    scores: List[float] = []
-    for i, sample in enumerate(datasets.Range(ds, args.num_samples)):
-        # Store the original question and answer for JSON output.
-        question_text = sample.audio_transcript
-        expected_answer = sample.messages[-1]["content"]
-        # Drop any assistant response from the sample.
-        sample.messages = sample.messages[:-1]
-        if not args.json:
-            run_tui(i, inference, sample, args, expected_answer, scores)
-        else:
-            output = inference.infer(
-                sample, max_tokens=args.max_tokens, temperature=args.temperature
+
+    if args.json:
+        dl = data_utils.DataLoader(
+            datasets.Range(ds, args.num_samples),
+            batch_size=args.batch_size,
+            collate_fn=lambda x: x,
+        )
+        current_batch = []
+        sample_index = 0
+        for current_batch in dl:
+            output = []
+            for sample in current_batch:
+                output.append(
+                    {
+                        "index": sample_index,
+                        "question": sample.audio_transcript,
+                        "expected_answer": sample.messages[-1]["content"],
+                    }
+                )
+                sample.messages = sample.messages[:-1]
+                sample_index += 1
+
+            output_batch = inference.infer_batch(
+                current_batch,
+                max_tokens=args.max_tokens,
+                temperature=args.temperature,
             )
-            obj = {
-                "question": question_text,
-                "generated_answer": output.text,
-                "expected_answer": expected_answer,
-            }
-            print(json.dumps(obj))
+            assert len(output) == len(output_batch)
+            for i, output_val in enumerate(output_batch):
+                output[i]["generated_answer"] = output_val.text
+            print(json.dumps(output))
```

Comment:
The `output` is a list, which is wrong I believe: each sample is assumed to be outputted on a different line.

Also here's a nitpicking proposal (needs to be combined with other change):
```suggestion
            assert len(current_batch) == len(output_batch)
            for sample, generated, expected  in zip(current_batch, output_batch, expected_answers):
                output = {
                    "index": sample_index,
                    "question": sample_in.audio_transcript,
                    "expected_answer": expected,
                    "generated_answer": generated.text,
                }
                sample_index += 1
                print(json.dumps(output))
```

Reply:
Thanks! Much cleaner :)